src/server/editorServices.ts

```diff
@@ -1217,13 +1217,35 @@ namespace ts.server {
             }
             else {
                 const project = this.createProject(configFilename, projectOptions);
+                let programSize = 0;
+
+                // As the project openning might not be complete if there are too many files,
+                // therefore to surface the diagnostics we need to make sure the given client file is opened.
+                if (clientFileName) {
+                    const currentClientFileInfo = this.openFile(clientFileName, /*openedByClient*/ true);
+                    project.addRoot(currentClientFileInfo);
+                    programSize += currentClientFileInfo.content.length;
+                }
+
                 for (const rootFilename of projectOptions.files) {
-                    if (this.host.fileExists(rootFilename)) {
-                        const info = this.openFile(rootFilename, /*openedByClient*/ clientFileName == rootFilename);
-                        project.addRoot(info);
+                    if (rootFilename === clientFileName) {
+                        continue;
+                    }
+
+                    if (programSize <= maxProgramSize) {

```

Comment:
I'm trying to understand why we need to repeat this check for the size of the content loaded here as well as in `createProgram`. It seems dangerous to have similar logic in both places they may behave slightly differently (or diverge unknowingly over time). What is the effect of **not** having this check here?